src/Microsoft.ML.Data/DataView/CompositeRowToRowMapper.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.ML.Runtime.Internal.Utilities;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    /// <summary>
+    /// A row-to-row mapper that is the result of a chained application of multiple mappers.
+    /// </summary>
+    public sealed class CompositeRowToRowMapper : IRowToRowMapper
+    {
+        private readonly IRowToRowMapper[] _innerMappers;
+        private static readonly IRowToRowMapper[] _empty = new IRowToRowMapper[0];
+
+        public ISchema InputSchema { get; }
+        public ISchema Schema { get; }
+
+        /// <summary>
+        /// Out of a series of mappers, construct a seemingly unitary mapper that is able to apply them in sequence.
+        /// </summary>
+        /// <param name="inputSchema">The input schema.</param>
+        /// <param name="mappers">The sequence of mappers to wrap. An empty or <c>null</c> argument
+        /// is legal, and counts as being a no-op application.</param>
+        public CompositeRowToRowMapper(ISchema inputSchema, IRowToRowMapper[] mappers)
+        {
+            Contracts.CheckValue(inputSchema, nameof(inputSchema));
+            Contracts.CheckValueOrNull(mappers);
+            _innerMappers = Utils.Size(mappers) > 0 ? mappers : _empty;
+            InputSchema = inputSchema;
+            Schema = Utils.Size(mappers) > 0 ? mappers[mappers.Length - 1].Schema : inputSchema;
+        }
+
+        public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+        {
+            Func<int, bool> toReturn = predicate;
+            for (int i = _innerMappers.Length - 1; i <= 0; --i)
+                toReturn = _innerMappers[i].GetDependencies(toReturn);
```

Comment:
>toReturn [](start = 16, length = 8)

does this look right?

Reply:
Looks right to me. We're chaining the delegates backwards.

---
In reply to: [219621916](https://github.com/dotnet/machinelearning/pull/973#discussion_r219621916) [](ancestors = 219621916)

Reply:
So the idea is that the predicates are (as elsewhere) predicates on the outputs, so if we are chaining the row-to-row mappers, then while we *transform* forwards, we have to check the dependencies in the opposite direction.

---
In reply to: [219622203](https://github.com/dotnet/machinelearning/pull/973#discussion_r219622203) [](ancestors = 219622203,219621916)